packages/grafana-e2e/src/flows/importDashboard.ts

```diff
@@ -0,0 +1,55 @@
+import { DeleteDashboardConfig } from '.';
+import { e2e } from '../index';
+import { fromBaseUrl, getDashboardUid } from '../support/url';
+
+type Panel = {
+  title: string;
+  [key: string]: any;
+};
+
+type Dashboard = { title: string; panels: Panel[]; [key: string]: any };
+
+/**
+ * Smoke test a datasource by quickly importing a test dashboard for it
+ * @param dashboardToImport a sample dashboard
+ */
+export const importDashboard = (dashboardToImport: Dashboard) => {
+  e2e().visit(fromBaseUrl('/dashboard/import'));
+
+  // Note: normally we'd use 'click' and then 'type' here, but the json object is so big that using 'val' is much faster
+  e2e.components.DashboardImportPage.textarea().invoke('val', JSON.stringify(dashboardToImport)).click({ force: true });
+  e2e.components.DashboardImportPage.submit().click({ force: true });
+  e2e.components.ImportDashboardForm.name().click({ force: true }).clear().type(dashboardToImport.title);
+  e2e.components.ImportDashboardForm.submit().click({ force: true });
+  e2e().wait(3000);
+
+  // save the newly imported dashboard to context so it'll get properly deleted later
+  e2e()
+    .url()
+    .should('contain', '/d/')
+    .then((url: string) => {
+      const uid = getDashboardUid(url);
+
+      e2e.getScenarioContext().then(({ addedDashboards }: any) => {
+        e2e.setScenarioContext({
+          addedDashboards: [...addedDashboards, { title: dashboardToImport.title, uid } as DeleteDashboardConfig],
+        });
+      });
+    });
+
+  // inspect first panel and verify data has been processed for it
+  e2e.components.Panels.Panel.title(dashboardToImport.panels[0].title).click({ force: true });
+  e2e.components.Panels.Panel.headerItems('Inspect').click({ force: true });
+  e2e.components.Tab.title('JSON').click({ force: true });
+  e2e().wait(3000);
+  e2e.components.PanelInspector.Json.content().contains('Panel JSON').click({ force: true });
+  e2e().wait(3000);
+  e2e.components.PanelInspector.Json.content().contains('Data').click({ force: true });
+  e2e().wait(3000);
+
+  // ensures that panel has loaded without knowingly hitting an error
+  // note: this does not prove that data came back as we expected it,
+  // it could get `state: Done` for no data for example
+  // but it ensures we didn't hit a 401 or 500 or something like that
+  e2e.components.CodeEditor.container().contains('"state": "Done"');
```

Comment:
not sure if relevant but maybe we can also check that the page doesn't render a `.panel-info-corner--error` ?

![Screenshot from 2021-07-07 11-06-42](https://user-images.githubusercontent.com/4025665/124732946-015acd00-df14-11eb-9e7b-3903b1ddcab5.png)

Reply:
yeah this was discussed a bit here: https://github.com/grafana/grafana/issues/33524

The problem is that if you test for the absence of something with cypress it will often give a false positive, because the page will not finish loading, it sees the error message wasn't there, and it then passes.

Vs when checking things that DO exists, cypress will automatically retry something a certain number of times by default before deciding the test has failed, in case the page is still loading. 

I'm pretty sure (though not 100%) that any state that causes the red error message to show the `state: "Error"` will also show. The one thing this does not test for is if we make a query and get back an empty values array.